### PR TITLE
added concurrency and other changes from stacked PRs

### PR DESCRIPTION
**Pull request by praveen-elastic**
Things included as a part of this PR:
- Multithreading
- Previous changes from other PRs

### CODE REVIEW

**Comment by artem-shelkovnikov** [comment]
@praveen-elastic can you rebase this PR on top of main, so that only the multithreading changes are present?

File: ees_sharepoint/fetch_index.py
Comment (artem-shelkovnikov) [inline ees_sharepoint/fetch_index.py:511]: I think it's worth to split this method into 2: 1 for list_items, another for drive_items. It'll make it easier to understand and handle - e.g. what if `job_type` passed is neither `list_items` nor `drive_items`, or situations like that.

**Comment by artem-shelkovnikov** [comment]
Hi @praveen-elastic, a couple notes to the review:

- ThreadPool can probably be re-used in all places it's used - logic is really similar, so might be possible to move all threading logic to a bunch of helper classes/methods, so that it would be easier to read the actual domain code behind them
- As we've discussed, the code in the threaded methods should ideally not use share memory, and I see that in the changes there's still a `library` dict object passed into the fetch methods, they should ideally be as pure and stateless as possible.

File: ees_sharepoint/fetch_index.py
Comment (moxarth-elastic) [reply ees_sharepoint/fetch_index.py:511]: Sure, we can split this function into two functions: one is for list_item and another is for drive_items.

File: ees_sharepoint/utils.py
Comment (moxarth-elastic) [reply ees_sharepoint/utils.py:35]: Sure, we’ll change the function name to `split_list_into_buckets`

File: ees_sharepoint/utils.py
Comment (moxarth-elastic) [reply ees_sharepoint/utils.py:64]: Primary use-case of this function is to have an approx. equal size of dictionaries to pass them in multiple threads for fetching the list_items and drive_items.

File: ees_sharepoint/utils.py
Comment (moxarth-elastic) [reply ees_sharepoint/utils.py:79]: Sure, we’ll change the function name to `split_date_range_into_chunks`.

File: ees_sharepoint/utils.py
Comment (moxarth-elastic) [reply ees_sharepoint/utils.py:94]: So as per your comment, I understood your confusion b/w these two functions so i decided to optimize it and merge them into one function. And documentation for the function is also updated to get more idea on what this function actually does.

**Comment by moxarth-elastic** [comment]
> Hi @praveen-elastic, a couple notes to the review:
> 
> * ThreadPool can probably be re-used in all places it's used - logic is really similar, so might be possible to move all threading logic to a bunch of helper classes/methods, so that it would be easier to read the actual domain code behind them
> * As we've discussed, the code in the threaded methods should ideally not use share memory, and I see that in the changes there's still a `library` dict object passed into the fetch methods, they should ideally be as pure and stateless as possible.

Sure, we’ll move the threading logic into a helper function. Also, we'll make all shared objects pure and stateless.

**Comment by tarekziade** [comment]
An event loop would work in a much more efficient way, and our WS client supports the pattern AFAIK. But using threads to run concurrent updates on WS will speed up the work since the app is I/O bound - so why not.

For the current patch, some general design notes:
* You are doing one request per document, this is very inefficient, and you are bombing WS
* You are creating thread pools within each function, this is quite complex
* There's nothing done to make sure we are thread-safe

I think the patch can be improved like this:

* create a documents Queue instance that will store data to index from the backend to the front end
* change the code that grabs documents so it pushes them into the queue
* change the code that index in WS so it pulls them from the queue in *batch* 
* create a pool of threads (=workers) for each side

Once you have this design, you can:

* tweak the batch size to reduce the number of queries made on WS
* tweak how many workers you want on each side (I doubt you need 40) 

Let me know if I can help

**Comment by praveen-elastic** [comment]
> An event loop would work in a much more efficient way, and our WS client supports the pattern AFAIK. But using threads to run concurrent updates on WS will speed up the work since the app is I/O bound - so why not.
> 
> For the current patch, some general design notes:
> 
> * You are doing one request per document, this is very inefficient, and you are bombing WS
> * You are creating thread pools within each function, this is quite complex
> * There's nothing done to make sure we are thread-safe
> 
> I think the patch can be improved like this:
> 
> * create a documents Queue instance that will store data to index from the backend to the front end
> * change the code that grabs documents so it pushes them into the queue
> * change the code that indexes in WS so it pulls them from the queue in _batch_
> * create a pool of threads (=workers) for each side
> 
> Once you have this design, you can:
> 
> * tweak the batch size to reduce the number of queries made on WS
> * tweak how many workers you want on each side (I doubt you need 40)
> 
> Let me know if I can help

@tarekziade  I have updated the multithreading approach based on our discussion i.e. adding a queue in between fetching from the SharePoint and indexing to the Enterprise Search. However, there's one outstanding issue in Windows that we are encountering : Since we are passing the class variables while initializing the Process, in the Windows OS, it throws Pickle error since it won't be able to pickle the class variables/objects. ref: https://github.com/elastic/enterprise-search-sharepoint-server-connector/blob/08e79e504db797f70d8283d1a73fdf9e4a102554/ees_sharepoint/full_sync_command.py#L31
You can start reviewing the rest of the changes and also let us know if you have some possible ideas on the issues

**Comment by praveen-elastic** [comment]
Hey, @tarekziade I've added a workaround to resolve pickle error in Windows OS: [commit link](https://github.com/elastic/enterprise-search-sharepoint-server-connector/pull/30/commits/2b60766df3af6a1f7f2341c72e6c40a362883f5a) Can you please take a look ?

**Comment by serenachou** [comment]
FYI @artem-shelkovnikov

File: ees_sharepoint/full_sync_command.py
Comment (tarekziade) [inline ees_sharepoint/full_sync_command.py:27]: Once you have a queue, producers and consumers are threads or processes, not just one.

The pattern to implement is in pseudo-code:

```
queue = Queue()

def producer():
    queue.put("data_from_3rd_party")

def consumer():
    batch = queue.get_100_lines()
    send_data_to_ws(batch)

producers = Pool(10)
consumers = Pool(10)

producers.start()
consumers.start()
```

The code that is fetching the data, and the code that is sending it to WS should have zero code related to threads or processes. It puts &  grabs data from a queue.

In an ideal world your code should be separated into two distinct parts (simple names here just to explain) :

- jobs.py: the queue code, and a function to start a pool of workers given a function
- fetch.py: the function that fetches the data from sharepoint
- sync.py : the function that sends it to WS 

The two last modules here should not know what a thread or a process is

**Comment by serenachou** [comment]
Please make sure @tarekziade  is good with this too based on our discussion today in our sync.

**Comment by serenachou** [comment]
@artem-shelkovnikov  @tarekziade  does this pass your final review for the multithreading?

**Comment by tarekziade** [comment]
@serenachou, there were no change done yet in the PR as discussed during the meeting. e.g. not having processes *and* threads and not mixing sharepoint code with jobs/workers handling

**Comment by praveen-elastic** [comment]
@artem-shelkovnikov, @tarekziade The PR is ready for review. Please take a look.

**Comment by artem-shelkovnikov** [comment]
Hi @praveen-elastic, as discussed before, please avoid making too many different changes under single PR - that makes it hard to review. This PR was supposed to add concurrency, and other changes like changes of quotes (' => "), renames of unrelated methods, changes in readme format should happen as separate requests.

I will review this PR, but for future I will ask you to split PRs into separate PRs per each individual functional change.

**Comment by praveen-elastic** [comment]
> Hi @praveen-elastic, as discussed before, please avoid making too many different changes under single PR - that makes it hard to review. This PR was supposed to add concurrency, and other changes like changes of quotes (' => "), renames of unrelated methods, changes in readme format should happen as separate requests.
> 
> I will review this PR, but for future I will ask you to split PRs into separate PRs per each individual functional change.

Sure, I will take a note of this.

File: ees_sharepoint/sync_sharepoint.py
Comment (artem-shelkovnikov) [inline ees_sharepoint/sync_sharepoint.py:94]: What if both are provided? Can we throw an error, if so?

File: ees_sharepoint/incremental_sync_command.py
Comment (artem-shelkovnikov) [inline ees_sharepoint/incremental_sync_command.py:83]: @tarekziade hi Tarek, do you think this approach is what we are looking for?

**Review by artem-shelkovnikov — CHANGES_REQUESTED** [review]
I've added small suggestions/questions, but mostly I'm waiting for review from @tarekziade regarding the parallelisation approach taken.

File: ees_sharepoint/sync_sharepoint.py
Comment (praveen-elastic) [reply ees_sharepoint/sync_sharepoint.py:94]: Hi @artem-shelkovnikov, In case of overlap between `include_fields` and `exclude_fields`, priority is given to `include_fields`.

File: ees_sharepoint/utils.py
Comment (praveen-elastic) [reply ees_sharepoint/utils.py]: Updated it to SharePoint

**Comment by serenachou** [comment]
@tarekziade  can you do one more review on this one? FYI @artem-shelkovnikov

**Review by tarekziade — APPROVED** [review]
Thanks for the refactoring; it addressed my concerns.

**Review by artem-shelkovnikov — APPROVED** [review]

**Pull request merged**
